system/mp/mpcomm.cpp

```diff
@@ -988,30 +1051,44 @@ protected: friend class CMPPacketReader;
                             e->Release();
                         }
                     }
-#ifdef _FULLTRACE
-                    PROGLOG("MP: rd = %d", rd);
-#endif
-                    /* NB: legacy clients that don't handle the exception deserialization here
-                     * will see reply as success, so no clean error,
-                     * but will fail shortly afterwards since server connection is closed
-                     */
-                    if (rd > sizeof(rd)) // legacy clients will only ever send a reply of 0 or 4, if greater, then new client is replying with an exception
+                } // while loopcnt
+
+                /* NB: legacy clients that don't handle the exception deserialization here
+                 * will see reply as success, so no clean error,
+                 * but will fail shortly afterwards since server connection is closed
+                 */
+                if (rd == sizeof(rd))
+                {
+                    int replyVal;
+                    memcpy(&replyVal, (int *)replyBuf, sizeof(int));
+                    if (replyVal > sizeof(ConnectHdr))
                     {
-                        MemoryBuffer mb;
-                        mb.setBuffer(rd, replyMem, false);
-                        size32_t len;
-                        mb.read(len); // exception length
-                        if (len)
+                        // read exception message ...
+                        MemoryBuffer replyMb;
+                        void *replyMem = replyMb.ensureCapacity(replyVal+1);
+
+                        size32_t xRead = 0;
+                        try
+                        {
+                            newsock->readtms(replyMem, replyVal, replyVal, xRead, CONNECT_READEXC_TIMEOUT);
+                        }
+                        catch (IException *e)
                         {
-                            exitException.setown(deserializeException(mb));
+                            if (0 == epStr.length())
+                            {
+                                SocketEndpoint ep;
+                                newsock->getPeerEndpoint(ep);
+                                ep.getUrlStr(epStr);
+                            }
+                            StringBuffer allowExcStr;
+                            exitException.setown(makeStringExceptionV(-99, "Error '%s' reading Allowlist exception from: %s", e->errorMessage(allowExcStr).str(), epStr.str()));
+                            e->Release();
                             throw exitException.getLink();
                         }
-                        break;
-                    }
-                    else if (rd != 0)
-                    {
-                        assertex(rd == sizeof(rd));
-                        break;
+                        MemoryBuffer mb;
```

Comment:
Should be able to deserialize diretcly from replyMb.  Possibly with a call to setLength() first.

Reply:
ok, fixed